pkg/n2k/service.go

```diff
@@ -0,0 +1,324 @@
+	"context"
+	"fmt"
+	"math"
+	"os"
+	"path"
+	"reflect"
+	"sync"
+	"time"
+
+	"github.com/boatkit-io/canbus"
+
+	"github.com/boatkit-io/tugboat/pkg/service"
+	"github.com/brutella/can"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+// Service is a management structure that implements a n2k canbus channel and
+// manages processing of canbus data into usable structures and events for the
+// rest of the system.
+//
+// This type implements the brutella.CanHandler interface.
+type Service struct {
+	log              *logrus.Logger
+	n2kInterfaceName string
+
+	shuttingDown bool
+	canbus       *canbus.Channel
+	pgnBuilder   *PGNBuilder
+
+	// Track subscriptions
+	subMutex   sync.Mutex
+	subs       map[SubscriptionId]*trackedSub
+	pgnSubs    map[string][]*trackedSub
+	globalSubs []*trackedSub
+	lastSubId  SubscriptionId
+}
+
+type SubscriptionId uint
+
+type trackedSub struct {
+	subId      SubscriptionId
+	structName string
+	// Will be either func(interface{}) for global handler or func(specific struct) for a struct callback
+	callback interface{}
+}
+
+var _ service.Activity = &Service{}
+
+// NewService returns a new instance of a n2k Service.
+func NewService(ctx context.Context, log *logrus.Logger, interfaceName string) (*Service, error) {
+	s := Service{
+		log:              log,
+		n2kInterfaceName: interfaceName,
+
+		shuttingDown: false,
+		canbus:       nil,
+
+		lastSubId:  0,
+		subs:       make(map[SubscriptionId]*trackedSub),
+		globalSubs: []*trackedSub{},
+		pgnSubs:    make(map[string][]*trackedSub),
+	}
+
+	s.pgnBuilder = NewPGNBuilder(log, s.handlePGNStruct)
+
+	return &s, nil
+}
+
+// Name returns the name of the Service service.Activity (helps implement the interface).
+func (*Service) Name() string {
+	return "n2k"
+}
+
+func (s *Service) Run(ctx context.Context) error {
+	if s.n2kInterfaceName != "" {
+		canbusOpts := canbus.ChannelOptions{
+			CanInterfaceName: s.n2kInterfaceName,
+			MessageHandler:   s.pgnBuilder.ProcessFrame,
+		}
+
+		if cc, err := canbus.NewChannel(ctx, s.log, canbusOpts); err != nil {
+			s.log.WithError(err).Warn("n2k channel creation failed")
+		} else {
+			s.canbus = cc
+		}
+	}
+
+	// Wait for shutdown
+	<-ctx.Done()
+
+	return nil
+}
+
+// Kill attempts to gracefully shutdown the Service.
+func (s *Service) Shutdown(ctx context.Context) error {
+	s.shuttingDown = true
+
+	var errs []error
+
+	if s.canbus != nil {
+		if err := s.canbus.Close(ctx); err != nil {
+			errs = append(errs, errors.Wrap(err, "closing n2k canbus channel"))
+		}
+	}
+
+	if len(errs) > 0 {
+		err := errs[0]
+		for i := 1; i < len(errs); i++ {
+			err = errors.Wrap(err, errs[i].Error())
+		}
+		return err
+	}
+	return nil
+}
+
+// Kill attempts to forcefully shutdown the Service.
+func (s *Service) Kill() error {
+	return nil
+}
+
+func (s *Service) ReplayFile(filename string) error {
+	wd, err := os.Getwd()
+	if err != nil {
+		return err
+	}
+
+	file, err := os.Open(path.Join(wd, "n2kreplays", filename))
```

Comment:
need to make this more generic -- make it take an absolute path, and we do the n2kreplays path joining in goatkit

Reply:
Deferring, since we'll move service.go out of public package.